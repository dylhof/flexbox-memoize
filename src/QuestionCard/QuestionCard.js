import React, { Component } from 'react';
import Question from '../Question/Question.js';
import Feedback from '../Feedback/Feedback.js';

class QuestionCard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      answerGiven: false,
      answerCorrect: false
    }
  }

  checkAnswer = (event) => {
    if(this.props.currentQuestion.correctAnswer === event.target.src) {
      this.setState({ answerCorrect: true });
      this.props.setCorrectAnswers();
      this.props.removeIncorrect(this.props.currentQuestion)
    } else { 
      this.setState({ answerCorrect: false })
      this.props.setIncorrect(this.props.currentQuestion);
    };
    this.setAnswerGiven();
  }
  
  setAnswerGiven = () => {
    this.setState({ answerGiven: true });
  }
  
  resetAnswerGiven = () => {
    this.setState({ answerGiven: false });
    this.props.setAnsweredQuestions();
  }

  render() {
    return(
    <div className='question-card'>   
      {!this.state.answerGiven && 
        <Question 
          currentQuestion={this.props.currentQuestion}
          checkAnswer={this.checkAnswer}
          randomAnswers={this.props.randomAnswers}/>
      }
      {this.state.answerGiven && 
        <Feedback 
          answerCorrect={this.state.answerCorrect}
          resetAnswerGiven={this.resetAnswerGiven}
          currentQuestion={this.props.currentQuestion}/>
      }
    </div>
    )
  }
}

export default QuestionCard;