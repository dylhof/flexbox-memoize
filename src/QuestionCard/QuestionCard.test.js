import React from 'react';
import QuestionCard from './QuestionCard';
import { shallow } from 'enzyme';

const currentQuestion = {
  "property": "flex-direction: row",
  "correctAnswer": "https://i.imgur.com/Xo1MD86.png",
  "answers": [
  "https://i.imgur.com/Xo1MD86.png",
  "https://i.imgur.com/JIJHnVf.png",
  "https://i.imgur.com/4y9cZDw.png",
  "https://i.imgur.com/Id0kLlR.png"
  ],
  "info": "This establishes the main-axis, thus defining the direction flex items are placed in the flex container. Using row organizes the items from left to right.",
  }
const answerClickMock = {target: {src: "https://i.imgur.com/Xo1MD86.png"}}
const wrongAnswerClickMock = {target: {src: "https://i.imgur.com/Id0kLlR.png"}}
const setCorrectAnswersMock = jest.fn();
const removeIncorrectMock = jest.fn();
const setIncorrectMock = jest.fn();
const setAnsweredQuestionsMock = jest.fn();

describe('QuestionCard', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <QuestionCard 
        currentQuestion={currentQuestion}
        setCorrectAnswers={setCorrectAnswersMock}
        removeIncorrect={removeIncorrectMock}
        setIncorrect={setIncorrectMock}
        setAnsweredQuestions={setAnsweredQuestionsMock}/>
    )
    wrapper.setState({answerCorrect: false, answerGiven: false})
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should check that answer is correct when passed the correct answer', () => {
    wrapper.instance().setAnswerGiven = jest.fn();
    expect(wrapper.state('answerCorrect')).toEqual(false);
    wrapper.instance().checkAnswer(answerClickMock);
    expect(wrapper.state('answerCorrect')).toEqual(true);
    expect(setCorrectAnswersMock).toBeCalled();
    expect(removeIncorrectMock).toBeCalled();
    expect(wrapper.instance().setAnswerGiven).toBeCalled();
  })
  
  it('should check that the answer is incorrect when passed an incorrect answer', () => {
    wrapper.instance().setAnswerGiven = jest.fn();
    wrapper.instance().checkAnswer(answerClickMock);
    expect(wrapper.state('answerCorrect')).toEqual(true);
    wrapper.instance().checkAnswer(wrongAnswerClickMock);
    expect(wrapper.state('answerCorrect')).toEqual(false);
    expect(setIncorrectMock).toBeCalled();
    expect(wrapper.instance().setAnswerGiven).toBeCalled();
  })

  it('should set the answer given state as true', () => {
    expect(wrapper.state('answerGiven')).toEqual(false);
    wrapper.instance().setAnswerGiven();
    expect(wrapper.state('answerGiven')).toEqual(true);
  })

  it('should reset the answerGiven state as false', () => {
    wrapper.instance().setAnswerGiven();
    expect(wrapper.state('answerGiven')).toEqual(true);
    wrapper.instance().resetAnswerGiven();
    expect(wrapper.state('answerGiven')).toEqual(false);
    expect(setAnsweredQuestionsMock).toBeCalled();
  })

})