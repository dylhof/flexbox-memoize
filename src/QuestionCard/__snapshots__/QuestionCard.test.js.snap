// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionCard should match the snapshot 1`] = `
<div
  className="question-card"
>
  <Question
    checkAnswer={[Function]}
    currentQuestion={
      Object {
        "answers": Array [
          "https://i.imgur.com/Xo1MD86.png",
          "https://i.imgur.com/JIJHnVf.png",
          "https://i.imgur.com/4y9cZDw.png",
          "https://i.imgur.com/Id0kLlR.png",
        ],
        "correctAnswer": "https://i.imgur.com/Xo1MD86.png",
        "info": "This establishes the main-axis, thus defining the direction flex items are placed in the flex container. Using row organizes the items from left to right.",
        "property": "flex-direction: row",
      }
    }
  />
</div>
`;
